### PR TITLE
ENH: Add Fiedler and fiedler companion to special matrices

### PR DESCRIPTION
**Pull request by ilayn**
These are in my opinion both interesting enough due to their properties and also once in a while useful to have. Especially the companion is better conditioned than the regular Frobenius companion which we might use for root finding when used with `matrix_balance` (which is also an active research area).

### CODE REVIEW

**Comment by ilayn** [comment]
In case there are no objections, I would like to merge this in a few days.

**Comment by ilayn** [comment]
@larsoner Can you help me understand what the problem is with the error here? After See Also is it tripping because it sees the first dot as a second reference ?

File: scipy/linalg/special_matrices.py [hunk 1]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
     elif scale == 'n':
         m /= n
     return m
+
+
+def fiedler(a):
+    """Returns a symmetric Fiedler matrix ``F[i, j] = np.abs(a[i] - a[j])``
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: Better to move the math to the next paragraph. I'm not sure this works.

File: scipy/linalg/special_matrices.py [hunk 2]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
...
+        coefficient array
+
+    Returns
+    -------
+    F : (n, n) ndarray
+
+    See Also
+    --------
+    circulant, toeplitz
+
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: remove extra blank line

File: scipy/linalg/special_matrices.py [hunk 3]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
...
+
+    .. versionadded:: 1.13.0
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: This belongs under `Notes`, that will be the issue you were seeing with the doc build

File: scipy/linalg/special_matrices.py [hunk 4]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
...
+
+    Given a polynomial coefficient array ``a``, this function forms a
+    pentadiagonal matrix with a special structure whose eigenvalues coincides
+    with the roots of ``a``.
+
+    Parameters
+    ----------
+    a : (N,) array_like
+        1-D array of polynomial coefficients in descending order with a nonzero
+        leading coefficient. For N < 2, an empty array is returned.
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: maybe put double backticks around `N < 2`

File: scipy/linalg/special_matrices.py [hunk 5]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
...
+
+    Returns
+    -------
+    c : (N-1, N-1) ndarray
+        Resulting companion matrix
+
+    Notes
+    -----
+    Similar to `scipy.linalg.companion` the leading coefficient should be
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: don't need `scipy.linalg`, `companion` will resolve

File: scipy/linalg/special_matrices.py [hunk 6]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
...
+    nonzero. In case the leading coefficient is not 1., other coefficients
+    are rescaled before the array generation. To avoid numerical issues, it is
+    best to provide a monic polynomial.
+
+    See Also
+    --------
+    companion
+
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: same here, missing `Notes` header

File: scipy/linalg/special_matrices.py [hunk 7]
```diff
@@ -1036,3 +1037,154 @@ def dft(n, scale=None):
...
+
+    .. versionadded:: 1.13.0
```
Comment (rgommers) [inline scipy/linalg/special_matrices.py]: 1.3.0 not 1.13.0

File: scipy/linalg/special_matrices.py (hunk 3 above)
Comment (rgommers) [reply scipy/linalg/special_matrices.py]: and 1.3.0 not 1.13.0

**Review by rgommers — CHANGES_REQUESTED** [review]
looks good, just some minor things to fix

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
Allright, all green and no more comments, in it goes. Thanks @ilayn

**Pull request merged**